### PR TITLE
Add tutorial for using SLAM for map building

### PR DESCRIPTION
**Pull request by mcdoerr**
Add a new tutorial on how to create custom maps of a local area using ROS's slam.

### CODE REVIEW

File: content/tutorials/mapping.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+title: "Using SLAM for Map Building"
+date: 2021-03-20T17:15:01-07:00
+summary: "Learn how to create custom maps of your car's surroundings"
```
Comment (schmittlema) [inline content/tutorials/mapping.md:4]: for some reason the apostrophe is coming up as some wierid symbols on the tutorials page. Maybe reword?

File: content/tutorials/mapping.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+difficulty: Intermediate
+duration: 0
+featured: true  # whether this is listed at / (must also be top 6 by weight)
+active: true    # whether this is listed at /tutorials/
+draft: false      # whether Hugo considers this a draft
+weight: 100
+---
+
```
Comment (schmittlema) [inline content/tutorials/mapping.md:12]: Need header image! Could be a example map, or a screenshot of rviz with the robot mapping

File: content/tutorials/mapping.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+## By: Madison Doerr and Markus Schiffer
+
+### Introduction
+The MuSHr car operates in a mapped area to determine its position, orientation, and plans. We can create
```
Comment (schmittlema) [inline content/tutorials/mapping.md:16]: Replace first sentence with: The MuSHR car uses a map of its environment to localize itself.

File: content/tutorials/mapping.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+active: true    # whether this is listed at /tutorials/
+draft: false      # whether Hugo considers this a draft
+weight: 100
+---
+
+## By: Madison Doerr and Markus Schiffer
+
+### Introduction
+The MuSHr car operates in a mapped area to determine its position, orientation, and plans. We can create 
+custom 2D maps for our cars by using [ROS's slam](http://wiki.ros.org/slam_toolbox) and our car's laser scanner to
```
Comment (schmittlema) [inline content/tutorials/mapping.md:17]: gmapping not ros slam: http://wiki.ros.org/gmapping

File: content/tutorials/mapping.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+### Goal
+Create a custom map of an area by driving the car to survey the surroundings.
+
+### Requirements
+  - Complete the [quickstart](/tutorials/quickstart) tutorial
+  - Complete the [first steps](/tutorials/first_steps) tutorial
+  - A desktop/laptop that can ssh into the car
+
+### Ssh into the car
+Boot up the car and connect the controller. Ssh into the car on your computer, replacing the IP address with your robot IP when connected with the ROBOT_AP network, or using the Wifi connect from the [first steps](/tutorials/first_steps) tutorial.
```
Comment (schmittlema) [inline content/tutorials/mapping.md:29]: bit confusing. I think the IP you have in the command is correct for ROBOT_AP, but not for the wifi network.

File: content/tutorials/mapping.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+```bash
+$ ssh robot@10.42.0.1 -X
+```
+
+### Install gmapping
+In a terminal sshed into the car using the above steps, make sure gmapping is installed.
```
Comment (schmittlema) [inline content/tutorials/mapping.md:35]: Was this installed for you by default?

File: content/tutorials/mapping.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+In a local terminal, ensure `ROS_IP` is set to your IP and `ROS_MASTER_URI` is set to the IP of the car. Then, in that terminal, run map_saver using the following command, replacing `map_name` with the desired name of the map.
+```bash
+$ rosrun map_server map_saver -f map_name map:=/car/map
+```
+This should create a .pgm file with the map and a .yaml file with the map metadata. 
+
+To load your new map into the simulator, move both files to `~/catkin_ws/src/mushr_sim/maps/` and edit `~/catkin_ws/src/mushr_sim/launch/map_server.launch` to set the map to the name of the .yaml file.
+
+### Touch up the map
+The map may have some stray pixels or jagged lines. To touch this up, we recommend using gimp to edit the .pgm file.
```
Comment (schmittlema) [inline content/tutorials/mapping.md:96]: link to gimp

File: content/tutorials/mapping.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+When doing this, use solid black pixels for edges, solid white areas for areas valid for the car, and solid gray areas
```
Comment (schmittlema) [inline content/tutorials/mapping.md:98]: Maybe give hex or rgb. Gray could be a lot of different things

File: content/tutorials/mapping.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+In a local terminal, ensure `ROS_IP` is set to your IP and `ROS_MASTER_URI` is set to the IP of the car. Then, in that terminal, run map_saver using the following command, replacing `map_name` with the desired name of the map.
+```bash
+$ rosrun map_server map_saver -f map_name map:=/car/map
+```
+This should create a .pgm file with the map and a .yaml file with the map metadata. 
+
+To load your new map into the simulator, move both files to `~/catkin_ws/src/mushr_sim/maps/` and edit `~/catkin_ws/src/mushr_sim/launch/map_server.launch` to set the map to the name of the .yaml file.
+
+### Touch up the map
```
Comment (schmittlema) [inline content/tutorials/mapping.md:95]: Touching

File: content/tutorials/mapping.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+### Save the map
+When you're satisfied with the map in rviz, you can save it to your computer using the following steps.
+
+In a local terminal, ensure `ROS_IP` is set to your IP and `ROS_MASTER_URI` is set to the IP of the car. Then, in that terminal, run map_saver using the following command, replacing `map_name` with the desired name of the map.
+```bash
+$ rosrun map_server map_saver -f map_name map:=/car/map
+```
+This should create a .pgm file with the map and a .yaml file with the map metadata. 
+
+To load your new map into the simulator, move both files to `~/catkin_ws/src/mushr_sim/maps/` and edit `~/catkin_ws/src/mushr_sim/launch/map_server.launch` to set the map to the name of the .yaml file.
```
Comment (schmittlema) [inline content/tutorials/mapping.md:93]: This seems like it is worth a separate section "Using the map" where you should where to save it for sim and for real and how to change the associated map_server files

**Review by schmittlema — CHANGES_REQUESTED** [review]
Great first draft! Read super clearly

File: content/tutorials/mapping.md (hunk 6 above)
Comment (mcdoerr) [reply content/tutorials/mapping.md:35]: Yeah, so this was installed by default, however, I could not get it to work without installing again and I think Markus had the same issue. Maybe we can switch this to sudo apt-update?

**Comment by mcdoerr** [comment]
Ok, should have addressed all the feedback! Thank you!

File: content/tutorials/mapping.md [hunk 11]
```diff
@@ -90,13 +98,40 @@ $ rosrun map_server map_saver -f map_name map:=/car/map
 ```
 This should create a .pgm file with the map and a .yaml file with the map metadata. 
 
-To load your new map into the simulator, move both files to `~/catkin_ws/src/mushr_sim/maps/` and edit `~/catkin_ws/src/mushr_sim/launch/map_server.launch` to set the map to the name of the .yaml file.
+### Using the map
```
Comment (schmittlema) [inline content/tutorials/mapping.md:101]: could you add a section for "In Sim" and "On Robot" because there are two map server files?

File: content/tutorials/mapping.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+Once you have sshed into the car, start teleop to start the laser scanner and control the car.
+```bash
+$ roslaunch mushr_base teleop.launch
+```
+Verify that the controller is working and you can move the car. If not, try reconnecting the controller and/or restarting teleop.
+
+In a separate terminal sshed into the robot, using the above process, start slam. Slam will take the data gathered from the laser 
+scanner started by teleop and transform that into a 2D map. 
+```bash
+$ roslaunch mushr_base slam.launch
```
Comment (schmittlema) [inline content/tutorials/mapping.md:64]: Put a note about making sure ROS_IP is set here. If it is not, the map won't show up in the next step

File: content/tutorials/mapping.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+When you're satisfied with the map in rviz, you can save it to your computer using the following steps.
+
+In a local terminal, ensure `ROS_IP` is set to your IP and `ROS_MASTER_URI` is set to the IP of the car. Then, in that terminal, run map_saver using the following command, replacing `map_name` with the desired name of the map.
+```bash
+$ rosrun map_server map_saver -f map_name map:=/car/map
+```
+This should create a .pgm file with the map and a .yaml file with the map metadata. 
+
+### Using the map
+#### On the sim
```
Comment (schmittlema) [inline content/tutorials/mapping.md:102]: Haven't tested yet

File: content/tutorials/mapping.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+$ export ROS_IP=YOUR-IP
+$ export ROS_MASTER_URI=http://ROBOT_IP:11311
+$ rviz
+```
+The map should appear, and you can move your robot into the map using `Set 2D Pos Estimate`.
+
+### Touching up the map
+The map may have some stray pixels or jagged lines. To touch this up, we recommend using [gimp](https://www.gimp.org/) to edit the .pgm file.
+
+When doing this, use solid black (hex: `#000000`) pixels for edges, solid white (hex: `#FFFFFF`) areas for areas valid for the car, and solid gray areas (hex: `#CDCDCD`) for invalid space for the car.
```
Comment (schmittlema) [inline content/tutorials/mapping.md:168]: My solid white showed up ast #FEFEFE

**Review by schmittlema — APPROVED** [review]
